include/Gaffer/Animation.h

```diff
@@ -381,6 +381,18 @@ class GAFFER_API Animation : public ComputeNode
 				/// Get the active key at the specified time, returns nullptr if no key with specified
 				/// time. (const access)
 				const Key *getKey( float time ) const;
+				/// Get the active in key, return nullptr if curve has no keys.
+				Key *getKeyIn();
+				/// Get the active in key, return nullptr if curve has no keys. (const access)
+				const Key *getKeyIn() const;
+				/// Get the active out key, return nullptr if curve has no keys.
+				Key *getKeyOut();
+				/// Get the active out key, return nullptr if curve has no keys. (const access)
+				const Key *getKeyOut() const;
+				/// Get the active key in the specified direction, return nullptr if curve has no keys.
+				Key *getKey( Animation::Direction direction );
+				/// Get the active key in the specified direction, return nullptr if curve has no keys. (const access)
+				const Key *getKey( Animation::Direction direction ) const;
```

Comment:
My preferences here aren't strong, but I think I found the old firstKey/finalKey naming clearer than keyIn/keyOut.

And having an overload of getKey that varies solely in argument type feels like it could occasionally be confusing.  I think I would prefer naming this getExtrapolationKey rather than just getKey.  I assume it wouldn't be useful in any context other than extrapolation?

Reply:
> My preferences here aren't strong, but I think I found the old firstKey/finalKey naming clearer than keyIn/keyOut.

I'd agree - and firstKey/lastKey would be even clearer. 

> And having an overload of getKey that varies solely in argument type feels like it could occasionally be confusing. 

I'd agree here too. Plus a `get` method always implies an associated `set` in Gaffer, which can't be the case here.

But most of all I don't think any of this needs to be in the public API.

Reply:
The functions are used in the gadget to implement drawing of the extrapolated curve so need to be public.

Reply:
If `get` implies `set` why is there a `getKey( time )` when there is no `setKey( time )` ?

Reply:
> If get implies set why is there a getKey( time ) when there is no setKey( time ) ?

Because I failed to notice the inconsistency :) My original thinking behind the Curve API for managing keys was that it should hold closely to the GraphComponent API for managing children (if Plugs weren't too heavyweight for storing lots of keys, then keys might even have been plugs). And that the time for a key is basically equivalent to the name of a plug (in that it's the property the parent indexes by). But whereas `setChild( "name", child )` felt reasonable and was useful, `setKey( time, key )` didn't, so I omitted it. But otherwise the same pattern holds (add, get, remove).

> The functions are used in the gadget to implement drawing of the extrapolated curve so need to be public.

Ah, sorry, I hadn't noticed that. I have to confess to doing a drive-by review where I just looked at the public API changes, leaving the rest to Daniel. I do feel that these overloads aren't helpful for clarity though, so I would propose simply `firstKey()` and `lastKey()`.

Reply:
how about `keyIn()` and `keyOut()` and `key( Direction)` ?

Reply:
> it would seem redundant to have two terminologies for the same concept

In my mind they're not the same concept. "First" and "Last" (or "front" and "back" in STL speak) are the endpoints of an ordered container. "In" and "Out" are directions, and come in a pair with nothing in between - appropriate for plugs and tangents but not for keys.

Reply:
ok, but it would seem there will be two terminologies, `curve.firstKey()` and `curve.key( Gaffer.Animation.Direction.In )` will be two ways to achieve the same thing using two different terminologies ? "first" and "in" ? similarly for "last" and "out". Are you suggesting another enum eg. `enum class KeyDirection { First, Last };` then having a function `curve.key( KeyDirection )` ? that seems more confusing to me.

Reply:
No, I'm suggesting that `firstKey()` and `lastKey()` are sufficient, just as `front()` and `back()` are. That may mean doing something like `k = direction == In ? firstKey() : lastKey()` in the AnimationEditor, but so be it.

Reply:
fair enough

Reply:
code now uses `firstKey()` and `lastKey()` naming.